### ISSUE TITLE
Training loop: multi-stage contrastive embedding training on MLX

### ISSUE THREAD

**Issue by pierretokns** [pierretokns/agenthub#5]
## Training Loop Design

Adapts the autoresearch-mlx autonomous loop for contrastive embedding training.

### Key differences from autoresearch-mlx

| Aspect | autoresearch-mlx | autoresearch-embed |
|--------|-----------------|-------------------|
| Objective | Next-token prediction | Contrastive learning (InfoNCE) |
| Model init | Random, train from scratch | Pretrained encoder, fine-tune |
| Output | Token logits | Normalized embedding vectors |
| Data format | Raw text sequences | (query, positive, negatives) triplets |
| Metric | Single scalar: val_bpb | Multi-dimensional: 8 MTEB scores |
| Time budget | Fixed 5 min | Variable: 55-67 min across stages |
| Training | Single loop | Multi-stage pipeline |
| Data | Fixed dataset | Dynamic (HF, synthetic, mined negatives) |

### Multi-stage pipeline per experiment

**Stage 1: Warmup (10 min)**
- Easy positive pairs (NLI entailment)
- InfoNCE, temperature 0.1, in-batch negatives only
- High LR (1e-4), linear warmup then constant
- Adapt projection head without destroying pretrained knowledge

**Stage 2: Full contrastive (20 min)**
- All curated datasets, configurable sampling ratios
- InfoNCE, temperature 0.05, in-batch + explicit negatives
- Lower LR (5e-5), cosine decay
- Main training phase

**Stage 3: Hard negative mining (5 min, inference only)**
- Embed full training corpus with current model
- Find top-k nearest non-positive passages per query
- Save assignments for Stage 4

**Stage 4: Hard negative fine-tuning (20 min)**
- Training data + mined hard negatives
- InfoNCE, temperature 0.02, weighted hard negatives
- Very low LR (1e-5), cosine decay
- Refine embedding space

**Total: ~55 min training + ~2 min quick eval + ~10 min full eval ≈ 67 min/experiment**
**Expected: 20-25 experiments per 24 hours**

### Base model options
- **ModernBERT-base (149M)** — proven embedding backbone, small and fast
- **Qwen3-0.6B** — mirrors pplx-embed approach, larger but more capable

### Memory estimates (M2 Ultra 64GB)
- ModernBERT + batch 256, seq 512: ~8GB peak
- Qwen3-0.6B + batch 64, seq 512: ~12GB peak
- Hard negative mining (1M passages): ~4GB embeddings + model
- Total peak: 15-20GB, well within budget
- Guard: abort if peak >55GB

### Experiment loop
```
REPEAT:
  1. Read state (results.tsv, leaderboard.tsv, git log)
  2. Pick ONE change (data > loss > schedule > architecture > hard negs)
  3. Edit src/ or configs/
  4. git add src/ configs/ && git commit -m "experiment: <desc>"
  5. uv run src/train.py > run.log 2>&1
  6. Handle crashes (3 retries max)
  7. Quick eval → decision gate
  8. If improved + clean contamination → full MTEB, keep, push, post
  9. If not improved → discard, reset
  10. Periodic: re-mine negatives every ~5 exp, re-search HF every ~10
```
